### PR TITLE
issue 303 STM quickstart

### PR DESCRIPTION
**Pull request by mmusgrov**
https://github.com/quarkusio/quarkus-quickstarts/issues/303

This quickstart shows how to use the Narayana STM extension added for quarkus issue [#4260](https://github.com/quarkusio/quarkus/issues/4260). The documentation for the extension is available in [extension guide](https://github.com/quarkusio/quarkus/blob/master/docs/src/main/asciidoc/stm-guide.adoc).

### CODE REVIEW

File: using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import io.restassured.RestAssured;
+import org.apache.http.HttpStatus;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+class STMResourceTest {
```
Comment (cescoffier) [inline using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java]: Can we have the @SubstrateTest equivalent?

File: using-stm/src/main/java/org/acme/quickstart/stm/FlightResource.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+@RequestScoped
+public class FlightResource {
+    @Inject
+    FlightServiceFactory factory;
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public CompletionStage<String> bookingCount() {
+        return CompletableFuture.supplyAsync(
+                () -> getInfo(factory.getInstance())
```
Comment (cescoffier) [inline using-stm/src/main/java/org/acme/quickstart/stm/FlightResource.java]: We don't recommend using `supplyAsync` without a specific _executor_. Can you adjust?

File: using-stm/src/main/java/org/acme/quickstart/stm/FlightResource.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        );
+    }
+
+    @POST
+    @Produces(MediaType.TEXT_PLAIN)
+    public CompletionStage<String> asynBook() {
+        return CompletableFuture.supplyAsync(() -> {
+            FlightService flightService = factory.getInstance();
```
Comment (cescoffier) [inline using-stm/src/main/java/org/acme/quickstart/stm/FlightResource.java]: Same comment here.

File: pom.xml [hunk 4]
```diff
@@ -45,5 +45,6 @@
         <module>dynamodb-client</module>
         <module>using-oauth2-rbac</module>
         <module>artemis-jms</module>
+        <module>using-stm</module>
```
Comment (cescoffier) [inline pom.xml]: I've found the name not necessarily self-descriptive. Could we come up with something a bit more expressive?

Comment (cescoffier) [reply pom.xml]: Something like `using-software-transactional-memory`.

File: using-stm/src/main/docker/Dockerfile.jvm [hunk 5]
```diff
@@ -0,0 +1,28 @@
+####
```
Comment (cescoffier) [inline using-stm/src/main/docker/Dockerfile.jvm]: I guess we can drop the 2 Dockerfiles.

File: using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Test
+    void testPost() {
+        String responseString;
+
+        makeBooking();
+        responseString = makeBooking();
+
+        System.out.printf("%s%n", responseString);
```
Comment (cescoffier) [inline using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java]: Can you remove the trace.

File: using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        assertThat(responseString, containsString("Booking Count=2"));
+    }
+
+    private String makeBooking() {
+        return RestAssured.post("/stm").then()
+                .assertThat()
+                .statusCode(HttpStatus.SC_OK)
```
Comment (cescoffier) [inline using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java]: You use `200` directly above.

File: using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java (hunk 1 above)
Comment (mmusgrov) [reply using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java]: +1

File: using-stm/src/main/java/org/acme/quickstart/stm/FlightResource.java (hunk 2 above)
Comment (mmusgrov) [reply using-stm/src/main/java/org/acme/quickstart/stm/FlightResource.java]: +1

File: using-stm/src/main/java/org/acme/quickstart/stm/FlightResource.java (hunk 3 above)
Comment (mmusgrov) [reply using-stm/src/main/java/org/acme/quickstart/stm/FlightResource.java]: +1

File: pom.xml (hunk 4 above)
Comment (mmusgrov) [reply pom.xml]: +1

File: using-stm/src/main/docker/Dockerfile.jvm (hunk 5 above)
Comment (mmusgrov) [reply using-stm/src/main/docker/Dockerfile.jvm]: +1

File: using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java (hunk 6 above)
Comment (mmusgrov) [reply using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java]: +1

File: using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java (hunk 7 above)
Comment (mmusgrov) [reply using-stm/src/test/java/org/acme/quickstart/stm/STMResourceTest.java]: +1

**Comment by mmusgrov** [comment]
Thanks for the review @cescoffier

**Comment by cescoffier** [comment]
Thanks!

**Pull request merged**